### PR TITLE
Add `CommitmentInput` and `BlockIssuanceCreditInput`

### LINKED ISSUE #901: Add CommitmentInput - rust
https://github.com/iotaledger/tips-draft/blob/tip45/tips/TIP-0045/tip-0045.md#commitment-input

### LINKED ISSUE #906: Add BlockIssuanceCreditInput - rust
https://github.com/iotaledger/tips-draft/blob/tip45/tips/TIP-0045/tip-0045.md#block-issuance-credit-input

### PR DESCRIPTION
**Pull request by abdulmth**
## Links to any relevant issues
fixes #901
fixes #906

## Type of change

Choose a type of change, and delete any options that are not relevant.
- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

File: sdk/src/types/block/context_input/mod.rs [hunk 1]
```diff
@@ -1,28 +1,42 @@
...
 pub use self::reward::RewardContextInput;
+
 use crate::types::block::Error;
+use derive_more::Display;
+use derive_more::From;
 
 /// A Context Input provides additional contextual information for the execution of a transaction, such as for different
 /// functionality related to accounts, commitments, or Mana rewards. A Context Input does not need to be unlocked.
-#[derive(Clone, Eq, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
+#[derive(Clone, Eq, Display, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
```
Comment (abdulmth) [inline sdk/src/types/block/context_input/mod.rs:18]: Not sure we need `Display` here.

Comment (abdulmth) [inline sdk/src/types/block/context_input/mod.rs:18]: Not sure we need `Display` here.

File: sdk/src/types/block/context_input/mod.rs [hunk 2]
```diff
@@ -1,28 +1,42 @@
...
 
-use derive_more::From;
-
+pub use self::block_issuance_credit::BlockIssuanceCreditContextInput;
+pub use self::commitment::CommitmentContextInput;
 pub use self::reward::RewardContextInput;
+
 use crate::types::block::Error;
+use derive_more::Display;
+use derive_more::From;
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: These goes after the mods

File: sdk/src/types/block/context_input/mod.rs [hunk 3]
```diff
@@ -1,28 +1,42 @@
...
 /// functionality related to accounts, commitments, or Mana rewards. A Context Input does not need to be unlocked.
-#[derive(Clone, Eq, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
+#[derive(Clone, Eq, Display, PartialEq, Hash, Ord, PartialOrd, From, packable::Packable)]
 #[packable(unpack_error = Error)]
 #[packable(tag_type = u8, with_error = Error::InvalidContextInputKind)]
 pub enum ContextInput {
+    /// A [`CommitmentContextInput`].
+    #[packable(tag = CommitmentContextInput::KIND)]
+    Commitment(CommitmentContextInput),
+
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: We don't usually put new lines between variants

File: sdk/src/types/block/context_input/mod.rs [hunk 4]
```diff
@@ -41,31 +57,69 @@ impl ContextInput {
...
 
+    /// Checks whether the context input is a [`CommitmentContextInput`].
+    pub fn is_commitment(&self) -> bool {
+        matches!(self, Self::Commitment(_))
+    }
+
+    /// Checks whether the context input is a [`BlockIssuanceCreditContextInput`].
+    pub fn is_block_issuance_credit(&self) -> bool {
+        matches!(self, Self::BlockIssuanceCredit(_))
+    }
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: Put them in the same order of the variants pls

File: sdk/src/types/block/context_input/mod.rs (hunk 1 above)
Comment (thibault-martinez) [reply sdk/src/types/block/context_input/mod.rs:18]: Yeah it's fine 👍🏻

File: sdk/src/types/block/context_input/mod.rs [hunk 5]
```diff
@@ -32,40 +44,80 @@ impl ContextInput {
...
         matches!(self, Self::Reward(_))
     }
 
+    /// Gets the input as an actual [`CommitmentContextInput`].
+    /// PANIC: do not call on a non-commitment context input.
+    pub fn as_commitment(&self) -> &CommitmentContextInput {
+        if let Self::Commitment(input) = self {
+            input
+        } else {
+            panic!("context input is not of type commitment: {:?}", self);
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs]: Please use the same panic message as for other enums https://github.com/iotaledger/iota-sdk/blob/2.0/sdk/src/types/block/output/mod.rs#L258

File: sdk/src/types/block/context_input/commitment.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use derive_more::{Display, From};
+
+use crate::types::block::slot::SlotCommitmentId;
+
+/// A Commitment Context indicates that the input references a commitment to a certain slot.
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/commitment.rs]: Just `Commitment Context` is a bit weird

File: sdk/src/types/block/context_input/commitment.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        self.0
+    }
+}
+
+pub(crate) mod dto {
+    use serde::{Deserialize, Serialize};
+
+    use super::*;
+
+    /// A Commitment Context indicates that the input references a commitment to a certain slot.
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/commitment.rs]: same

File: sdk/src/types/block/context_input/mod.rs [hunk 8]
```diff
@@ -32,40 +44,80 @@ impl ContextInput {
...
     pub fn kind(&self) -> u8 {
         match self {
+            Self::Commitment(_) => CommitmentContextInput::KIND,
+            Self::BlockIssuanceCredit(_) => BlockIssuanceCreditContextInput::KIND,
             Self::Reward(_) => RewardContextInput::KIND,
         }
     }
 
+    /// Checks whether the context input is a [`CommitmentContextInput`].
+    pub fn is_commitment(&self) -> bool {
```
Comment (thibault-martinez) [inline sdk/src/types/block/context_input/mod.rs:55]: Ok I'm super annoying now but in other modules we alternate the is_ and the as_ methods 😬

File: sdk/src/types/block/context_input/commitment.rs (hunk 6 above)
Comment (abdulmth) [reply sdk/src/types/block/context_input/commitment.rs]: from the TIP: `A Commitment Input is an input that allows to reference a commitment to a certain slot` is it better? I didn't like `input is an input` so I changed it a bit.

File: sdk/src/types/block/context_input/mod.rs (hunk 8 above)
Comment (abdulmth) [reply sdk/src/types/block/context_input/mod.rs:55]: Not sure what do you mean, for me the `is_` should return a `bool` and the `as_` should  cast, which is what we do here 🤔

Comment (thibault-martinez) [reply sdk/src/types/block/context_input/mod.rs:55]: I mean in the order. We do is_ then as_ for each variant

File: sdk/src/types/block/context_input/commitment.rs (hunk 6 above)
Comment (thibault-martinez) [reply sdk/src/types/block/context_input/commitment.rs]: I would either just say `Commitment` or `Commitment Input`

File: sdk/src/types/block/context_input/mod.rs (hunk 8 above)
Comment (abdulmth) [reply sdk/src/types/block/context_input/mod.rs:55]: 😆 Ok I'll change it

**Review by thibault-martinez — DISMISSED** [review]

**Review by qrayven — DISMISSED** [review]

File: sdk/src/types/block/context_input/block_issuance_credit.rs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            Self {
+                kind: BlockIssuanceCreditContextInput::KIND,
+                account_id: value.account_id(),
+            }
+        }
+    }
+
+    impl From<BlockIssuanceCreditContextInputDto> for BlockIssuanceCreditContextInput {
+        fn from(value: BlockIssuanceCreditContextInputDto) -> Self {
+            Self::new(value.account_id)
```
Comment (kwek20) [inline sdk/src/types/block/context_input/block_issuance_credit.rs:56]: check if kind is 1¿ I think we did that in thoralfs pr

Comment (abdulmth) [reply sdk/src/types/block/context_input/block_issuance_credit.rs:56]: this is what Thoralf had: 
```rust
    impl From<RewardContextInputDto> for RewardContextInput {
        fn from(value: RewardContextInputDto) -> Self {
            Self::new(value.index)
        }
    }
``` 

do you suggest to implement `TryFrom` instead? 🤔

Comment (kwek20) [reply sdk/src/types/block/context_input/block_issuance_credit.rs:56]: huh leme see where i saw that what i refer to

Comment (kwek20) [reply sdk/src/types/block/context_input/block_issuance_credit.rs:56]: ah never mind was talking about custom ser/de

**Review by kwek20 — DISMISSED** [review]

File: sdk/src/types/block/context_input/block_issuance_credit.rs [hunk 10]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use derive_more::{Display, From};
+
+use crate::types::block::output::AccountId;
+
+/// A Block Issuance Credit Input provides the VM with context for the value of
+/// the BIC vector of a specific slot.
```
Comment (DaughterOfMars) [inline sdk/src/types/block/context_input/block_issuance_credit.rs:9]: Can we write out BIC?

Comment (DaughterOfMars) [reply sdk/src/types/block/context_input/block_issuance_credit.rs:9]: or maybe rephrase to avoid repetition, I dunno. Maybe put (BIC) after the first one.

**Review by DaughterOfMars — CHANGES_REQUESTED** [review]
Now that the DTO PR is merged, this should be refactored to match.
1. The `dto` modules should be private.
2. There should be `impl_typed_serde_dto!` calls for each variant dto
3. The enum should derive serde and be `untagged`

File: sdk/src/types/block/context_input/block_issuance_credit.rs [hunk 11]
```diff
@@ -5,7 +5,7 @@ use derive_more::{Display, From};
 
 use crate::types::block::output::AccountId;
 
-/// A Block Issuance Credit Input provides the VM with context for the value of
+/// A Block Issuance Credit Input (BIC Input) provides the VM with context for the value of
```
Comment (DaughterOfMars) [inline sdk/src/types/block/context_input/block_issuance_credit.rs]:
```suggestion
/// A Block Issuance Credit (BIC) Input provides the VM with context for the value of
```

File: sdk/src/types/block/context_input/reward.rs [hunk 12]
```diff
@@ -49,4 +49,6 @@ pub(crate) mod dto {
             Self::new(value.index)
         }
     }
+
+    impl_serde_typed_dto!(RewardContextInput, RewardContextInputDto, "reward input");
```
Comment (DaughterOfMars) [inline sdk/src/types/block/context_input/reward.rs]:
```suggestion
    impl_serde_typed_dto!(RewardContextInput, RewardContextInputDto, "reward context input");
```

File: sdk/src/types/block/context_input/commitment.rs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        }
+    }
+
+    impl From<CommitmentContextInputDto> for CommitmentContextInput {
+        fn from(value: CommitmentContextInputDto) -> Self {
+            Self::new(value.commitment_id)
+        }
+    }
+    impl_serde_typed_dto!(CommitmentContextInput, CommitmentContextInputDto, "commitment input");
```
Comment (DaughterOfMars) [inline sdk/src/types/block/context_input/commitment.rs]:
```suggestion
    impl_serde_typed_dto!(CommitmentContextInput, CommitmentContextInputDto, "commitment context input");
```

File: sdk/src/types/block/context_input/block_issuance_credit.rs (hunk 11 above)
Comment (abdulmth) [reply sdk/src/types/block/context_input/block_issuance_credit.rs]: It didn't sound right to put the abbreviation in the middle 🤔

File: sdk/src/types/block/context_input/block_issuance_credit.rs [hunk 14]
```diff
@@ -26,6 +26,8 @@ impl BlockIssuanceCreditContextInput {
 }
 
 mod dto {
+    use alloc::format;
```
Comment (abdulmth) [inline sdk/src/types/block/context_input/block_issuance_credit.rs:29]: Couldn't this be imported in the macro? 🤔

Comment (qrayven) [reply sdk/src/types/block/context_input/block_issuance_credit.rs:29]: I actually I fixed that In my PR #984 :)

File: sdk/src/types/block/context_input/block_issuance_credit.rs [hunk 15]
```diff
@@ -53,4 +54,10 @@ pub(crate) mod dto {
             Self::new(value.account_id)
         }
     }
+
+    impl_serde_typed_dto!(
+        BlockIssuanceCreditContextInput,
+        BlockIssuanceCreditContextInputDto,
+        "block issuance credit input"
```
Comment (Thoralf-M) [inline sdk/src/types/block/context_input/block_issuance_credit.rs]:
```suggestion
        "block issuance credit context input"
```

Comment (abdulmth) [reply sdk/src/types/block/context_input/block_issuance_credit.rs]: This is already done, isn't it? or GitHub is buggy again?

Comment (Thoralf-M) [reply sdk/src/types/block/context_input/block_issuance_credit.rs]: Yeah now I also see it updated

File: sdk/src/types/block/context_input/block_issuance_credit.rs (hunk 11 above)
Comment (kwek20) [reply sdk/src/types/block/context_input/block_issuance_credit.rs]: Its common to but the abbreviation behind the words though, i approve of this suggestion :=)

**Review by DaughterOfMars — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Review by qrayven — APPROVED** [review]
